src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsGroupComponent.scala

```diff
@@ -1,44 +1,177 @@
+      val subgroupInserts = rawlsGroup.subGroups.toSeq.map {
+        groupSubgroupsQuery insertOrUpdate marshalGroupSubgroups(_, rawlsGroup)
+      }
+
+      val actions = groupInsert andThen userDeletes andThen DBIO.seq(userInserts: _*) andThen subGroupDeletes andThen DBIO.seq(subgroupInserts: _*)
+      actions map { _ => rawlsGroup }
+    }
+
+    def load(groupRef: RawlsGroupRef): ReadAction[Option[RawlsGroup]] = {
+      val name = groupRef.groupName.value
+      val groupID = s"group with name '$name'"
+      loadGroup(findGroupByName(name), groupID)
+    }
+
+    def loadGroupByEmail(groupEmail: RawlsGroupEmail): ReadAction[Option[RawlsGroup]] = {
+      val email = groupEmail.value
+      val groupID = s"group with email '$email'"
+      loadGroup(findGroupByEmail(email), groupID)
+    }
+
+    def delete(groupRef: RawlsGroupRef): ReadWriteAction[Boolean] = {
+      val name = groupRef.groupName.value
+      val groupQuery = findGroupByName(name)
+
+      uniqueResult[RawlsGroupRecord](groupQuery).flatMap {
+        case None => DBIO.successful(false)
+        case Some(groupRec) =>
+          val actions = findUsersByGroupName(name).delete andThen findSubgroupsByGroupName(name).delete andThen groupQuery.delete
+          actions map { count => count > 0 }
+      }
+    }
+
+    def loadFromEmail(email: String): ReadAction[Option[Either[RawlsUser, RawlsGroup]]] = {
+      val user: ReadAction[Option[RawlsUser]] = rawlsUserQuery.loadUserByEmail(RawlsUserEmail(email))
+      val group: ReadAction[Option[RawlsGroup]] = loadGroupByEmail(RawlsGroupEmail(email))
+
+      user zip group map {
+        case (Some(u), Some(g)) => throw new RawlsException(s"Database error: email $email refers to both a user and a group")
+        case (Some(u), None) => Option(Left(u))
+        case (None, Some(g)) => Option(Right(g))
+        case _ => None
+      }
     }
+
+    def loadGroupIfMember(groupRef: RawlsGroupRef, userRef: RawlsUserRef): ReadAction[Option[RawlsGroup]] = {
+      checkMembershipRecursively(userRef, Set.empty, Set(groupRef)).flatMap {
+        case true => load(groupRef)
+        case false => DBIO.successful(None)
+      }
+    }
+
+    private def checkMembershipRecursively(userRef: RawlsUserRef, previouslyCheckedGroups: Set[RawlsGroupRef], groupsToCheck: Set[RawlsGroupRef]): ReadAction[Boolean] = {
+      groupsToCheck.headOption match {
+        case Some(checkingNow) =>
+          val remainingToCheck = groupsToCheck - checkingNow
+
+          load(checkingNow).flatMap {
+            case Some(group) if group.users.contains(userRef) => DBIO.successful(true)
+            case Some(group) =>
+              val newGroups = group.subGroups -- previouslyCheckedGroups
+              checkMembershipRecursively(userRef, previouslyCheckedGroups + checkingNow, remainingToCheck ++ newGroups)
+            case _ => checkMembershipRecursively(userRef, previouslyCheckedGroups + checkingNow, remainingToCheck)
```

Comment:
is this the case where we try to load a group and it does not exist? This would indicate corrupt data, correct? I am fine with being tolerant of that here, but add a comment.